### ISSUE TITLE
Retrieve the latest configuration before sending an email

### ISSUE THREAD

**Issue by schmitch** [playframework/play-mailer#79]
Hello, currently we have something like that:

```scala
class SMTPMailer @Inject() (smtpConfiguration: SMTPConfiguration) extends MailerClient {
```

However the Provider doesn't help here since the SMTPMailer will just do the same as before.

Currently it would be great if SMTPMailer stays the same so that we contain compability, however for the injection point it would be better to have something like:
```scala
class ConfigurableMailer @Inject()(smtpConfigurationProvider: Provider[SMTPConfiguration]) extends MailerClient {

  override def send(data: Email): String = {
    new SMTPMailer(smtpConfigurationProvider.get()).send(data)
  }

}
```

If we don't the Provider makes just no sense.

**Comment by Mogztter** [comment]
Hello @schmitch, the current implementation is designed to work with a configuration defined in the `application.conf` and never changed. In this case this make sense to have exactly one instance of SMTP configuration.

In your example, you will create a SMTP configuration every time you send an email. This can be useful if you read your configuration from a database and want to make sure that you always have the "latest" configuration.

I think both uses cases are valid and we need to provide a solution for both of them.

I think this issue should be renamed "Retrieve the latest configuration before sending an email", wdyt ?

**Comment by schmitch** [comment]
@Mogztter Hm. I'm not sure if there is "too much" overhead if we retrieve a configuration object everytime we send a mail. 
I mean we would need to change the Provider slightly so that it uses a singleton pattern, too so:

we would have:
```
class SMTPConfigurationProvider @Inject()(smtpConfiguration: SMTPConfiguration) extends Provider[SMTPConfiguration] {
  override def get() = {
    smtpConfiguration
  }
}
```

Now the Provider would use the SMTPConfiguration. However it is configurable since we could overwrite the provider if combined with the src from above. The only overhead it would produce would be some more function calls.

**Comment by Mogztter** [comment]
> I'm not sure if there is "too much" overhead if we retrieve a configuration object everytime we send a mail. 
> The only overhead it would produce would be some more function calls.

We also creating a new `SMTPMailer` instance each time, which is necessary if the configuration is static but not sure if this is such a big deal...
My second concern is to provide a clean constructor to do compile time DI. Today we can write something as simple as:

```scala
new SMTPMailer(SMTPConfiguration("typesafe.org", 1234)).send(email)
```

If we replace the constructor argument with `smtpConfigurationProvider: Provider[SMTPConfiguration]`, we will have to instantiate a Provider (Guice) and that's exactly what users want to avoid when they do compile time DI.

Having said that, I think that injecting Provider will bring flexibility when using runtime DI with a "dynamic" configuration. So we should really consider it.

> Btw. if you liked it I would provide a PR since I already wrote the source here anyway..

Yes of course :+1:  ! We definitely need to do a code review to make sure that we are covering all the use cases.

Thanks

**Comment by mkurz** [comment]
Any news on this?

**Comment by schmitch** [comment]
Oh yeah sorry I've totally forgotten about this, I will write some code today and push it

**Comment by mkurz** [comment]
@schmitch Thanks!

**Comment by schmitch** [comment]
I added #81 as a WIP PR however I found some duplicated things.

**Comment by schmitch** [comment]
Closed due to #81
